impeller/fixtures/impeller.frag

```diff
@@ -0,0 +1,436 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+uniform samplerCube cube_map;
+uniform sampler2D blue_noise;
+
+uniform FrameInfo {
+  vec2 texture_size;
+  float time;
+}
+frame_info;
+
+in vec2 v_screen_position;
+out vec4 frag_color;
+
+const float kPi = acos(-1.0);
+const float kHalfSqrtTwo = sqrt(2.0) / 2.0;
+const float kEpsilon = 0.001;
+
+// Materials (Albedo + reflectivity)
+const vec4 kBottomColor = vec4(0.0, 0.34, 0.61, 0.5);
+const vec4 kMiddleColor = vec4(0.16, 0.71, 0.96, 0.5);
+const vec4 kTopColor = vec4(0.33, 0.77, 0.97, 0.5);
+const vec4 kImpellerOuterColor = vec4(0.16, 0.71, 0.96, 0.5);
+const vec4 kImpellerRimColor = vec4(0.1, 0.1, 0.1, 0.0);
+const vec4 kImpellerBladeColor = vec4(0.1, 0.1, 0.1, 1.3);
+
+// Scene
+const int kMaxSteps = 70;
+const float kMaxDistance = 300.0;
+const vec3 kSunDirection = normalize(vec3(2, -5, 3));
+const float kGlowBlend = 1.1;
+const vec4 kGlowColor = vec4(0.86, 0.98, 1.0, 1);
+const vec4 kGlowColor2 = vec4(1.66, 0.98, 0.5, 1);
+const float kGlassIOR = 1.10;
+
+// Camera
+const float kFocalLength = 12.0;
+const float kApertureSize = 0.5;
+const int kRaysPerFrag = 4;
+
+mat3 RotateEuler(vec3 r) {
+  return mat3(
+      cos(r.x) * cos(r.y), cos(r.x) * sin(r.y) * sin(r.z) - sin(r.x) * cos(r.z),
+      cos(r.x) * sin(r.y) * cos(r.z) + sin(r.x) * sin(r.z), sin(r.x) * cos(r.y),
+      sin(r.x) * sin(r.y) * sin(r.z) + cos(r.x) * cos(r.z),
+      sin(r.x) * sin(r.y) * cos(r.z) - cos(r.x) * sin(r.z), -sin(r.y),
+      cos(r.y) * sin(r.z), cos(r.y) * cos(r.z));
+}
+
+//------------------------------------------------------------------------------
+/// Noise functions.
+///
+
+vec2 Hash(float seed) {
+  vec2 n = vec2(dot(vec2(seed, -0.1), vec2(13.8767971, 22.2091485)),
+                dot(vec2(seed, -0.2), vec2(12.3432217, 48.0579381)));
+  return fract(sin(n) * 24791.8159993);
+}
+
+vec4 BlueNoise(vec2 uv) {
+  // Remove mod once repeat sampling works.
+  return texture(blue_noise, mod(uv, 1.0));
+}
+
+vec4 BlueNoiseWithRandomOffset(vec2 screen_position, float seed) {
+  return BlueNoise(screen_position / 256.0 + Hash(seed));
+}
+
+//------------------------------------------------------------------------------
+/// Primitive distance functions.
+///
+
+float SphereDistance(vec3 sample_position,
```

Comment:
We should consider putting these in a header file with other useful utilities that we can import as needed. Our shaders so far have been extremely simple so it sort of didn't matter yet.

Reply:
So I've been thinking: Should we have a standard shader library that ships with impellerc itself and slowly graduate routines that get repeated use to it? Basically a directory that's always guaranteed to be in the include path and defines "namespaced" functions:
```glsl
#include "impeller/distance.glsl"

...

float my_sphere = ImSphereDistance(space, vec3(0), 1.0);
```

Perhaps worth discussing over a bug (and not sure if this is the best naming convention :)). A few other things that are good candidates: Sampler mode emulation for legacy platforms (i.e. ClampToBorder), common sigmoids, erf, gaussian/normal distribution, color space conversions, cheap hash/noise functions, simplex noise.

Reply:
The need for something like this is coming up in my work to address differences in coordinate spaces between OpenGL and Metal. https://docs.unity3d.com/Manual/SL-PlatformDifferences.html is probably recommended reading for inspiration as it documents some of the patterns you are thinking of as well as built-ins.

In any case, we can handle this at a later time. Just thought we might be at a point where our shaders are not so simple anymore.